### ISSUE TITLE
Support React 17

### ISSUE THREAD

**Issue by mquandalle** [diegomura/react-pdf#1062]
NPM v7 throws the following error when using react-pdf with the latest version of React 

```
npm ERR! node_modules/react
npm ERR!   react@"^17.0.0" from the root project
npm ERR! 
npm ERR! Could not resolve dependency:
npm ERR! peer react@"^16.0.0" from @react-pdf/renderer@1.6.12
npm ERR! node_modules/@react-pdf/renderer
npm ERR!   @react-pdf/renderer@"^1.6.10" from the root project
```

I'm not sure if there are any code migration required or if only the `package.json` needs to me modified.

**Comment by arobert93** [comment]
Any update on this?

**Comment by xavierraffin** [comment]
Interested too

**Comment by arobert93** [comment]
> it might help you `npm install @react-pdf/renderer --legacy-peer-deps`

This is a quick fix that solves the situation temporary, but if you remove the lock file, you have to use it again. It is not a clean solution...

**Comment by princefishthrower** [comment]
Made a pull request to bump react version here: https://github.com/diegomura/react-pdf/pull/1157
You can meanwhile just fork this and issue `npm update react`, it works fine and then all tests pass, not sure where any maintainers are on this one, the original React 17 version issue was raised last October...

**Comment by diegomura** [comment]
Hi! Sorry for the delay!
What would happen to projects using React 16?

**Comment by arobert93** [comment]
@diegomura Maybe you can add both, with version range, until you test it.

**Comment by diegomura** [comment]
So I tried upgrading react's peer dependency from `^16.8.6` to `>16.8.6` that matches new React 17 version, but NPM v7 stills throws peer dependency errors due to `react-reconciler@0.23.0` which has `react@16` as peer dependency as well... 

The newest `react-reconciler@0.26.2` has `react@^17.0.0` as peer dependency, but first it breaks react-pdf (still have to see why) and second, if I'm not mistaken will break react-pdf to anyone using `react@16`, which is still a lot of people.

It seems that the cause of this is NPMv7 now [installs peer dependencies](https://github.com/npm/rfcs/blob/latest/implemented/0025-install-peer-deps.md) and breaks things (this is apparently breaking the react library ecosystem, not just this package). I'm open to suggestions here but I do not see I can keep supporting all react versions and making NPMv7 to work, unless you use the `--legacy-peer-dep` flag as recommended above

**Comment by diegomura** [comment]
Just published `1.6.16` with the change suggested by `princefishthrower` above, but I think it's not enough to make this whole peerDependency check to work. Please try it and let me know!

**Comment by maxdev** [comment]
@diegomura I tried on 2.0.0 and 1.6.16, but unfotunately it doesn't work without --legacy-peer-dep flag

**Comment by diegomura** [comment]
Yes. As explained above, I couldn't find the way of keep supporting react 16 and 17. This whole peer dependency check in nvm@7 breaks this library. The issue is not just react-pdf peer dependency but specially `react-reconciler` peer dependency as well. If I define both react-reconciler versions on our side, npm still does not infer the correct version it has to use. I'm open to suggestions here

**Comment by maxdev** [comment]
Hi @diegomura 

I see the problem. Maybe solution with aliases could help in this case? https://dev.to/joshx/test-your-npm-package-against-multiple-versions-of-its-peer-dependency-34j4

**Comment by diegomura** [comment]
Interesting.. But I don't think npm aliases can solve this. I will still have to list a `react-reconciler` version with a peer dependency of react@16, so the check will still fail, right?. I thought that by defining `"react-reconciler: ^0.23.0 || ^0.26.0"` npm would automatically choose the version who's peer dependencies best adjusts to your dependencies (0.23.0 having react@16 and 0.26.0 react@17 respectively). But no, it always end up installing 0.26.0, breaking all those projects that uses react@16. 

React official renderers builds, such as react-dom or react-native, are shipped with the `react-reconciler` code embedded in the build (I mean, not as an external dependency). This is something I guess I can do for this project as well, although it just feels wrong to publish this package with a third-party dependency build on it

**Comment by maxdev** [comment]
Hi @diegomura 

> React official renderers builds, such as react-dom or react-native, are shipped with the react-reconciler code embedded in the build (I mean, not as an external dependency). This is something I guess I can do for this project as well, although it just feels wrong to publish this package with a third-party dependency built inside of it

I think it's a good point.

> Interesting.. But I don't think npm aliases can solve this. I will still have to list a react-reconciler version with a peer dependency of react@16, so the check will still fail, right?. I thought that by defining "react-reconciler: ^0.23.0 || ^0.26.0" npm would automatically choose the version who's peer dependencies best adjusts to your dependencies (0.23.0 having react@16 and 0.26.0 react@17 respectively). But no, it always end up installing 0.26.0, breaking all those projects that uses react@16.

Maybe I'm completely wrong, but I have an assumption why npm always installs latest minor version, probably because `^0.23.0` is equivalent to `^0.26.0`? 
What if  to try something like `"react-reconciler: 0.23.0 || ^0.26.0"`

**Comment by diegomura** [comment]
Thanks @maxdev ! That might be right. Will try today and report back here

**Comment by diegomura** [comment]
Unfortunately this still happens :(

![Screen Shot 2021-04-08 at 3 04 20 PM](https://user-images.githubusercontent.com/5600341/114075239-b8936980-987b-11eb-96e2-aea48b98555b.png)

npm keeps insisting on installing `react-reconciler@0.26.0` even if `react@16` is installed

**Comment by KarlBaumann** [comment]
I can confirm that it is still broken

**Comment by diegomura** [comment]
Thanks @KarlBaumann. It is :( but as explained above I'm on a dead end for both supporting react 16 and 17 with the new npm peer dependency install

**Comment by KarlBaumann** [comment]
@diegomura 
I would say you don't have to support outdated libraries in latest versions. If people want to use outdated software, they can use older releases.

**Comment by diegomura** [comment]
While I generally agree, with React for me it's more complex. Since it's the **main** peer dependency of this project and one of those libraries that people usually do not update right away, just stopping support for React 16 would leave many people behind. It's not that people using React17 cannot use latest build either. For what I understood, the only issue is with the latest npm peer dependency check, right?

**Comment by KarlBaumann** [comment]
for me it does not work also with the peer-dependency-check disabled

**Comment by mdodge-ecgrow** [comment]
Just curious, how long do you forsee supporting React 16?

**Comment by diegomura** [comment]
I don't have any fixed plan. I'm open to suggestions 😄  Would still love to see if there's some solution to this. @KarlBaumann releasing a 3.x it's not an option since that would be even more confusing

**Comment by mquandalle** [comment]
Is it possible to use a “or” condition ?

```
    "react": "^16.0.0 || ^17.0.0",
    "react-dom": "^16.0.0 || ^17.0.0",
```

**Comment by IsaacTrevino** [comment]
+1

**Comment by Sinakhx** [comment]
alson looking for react-17 support

**Comment by grmatthews** [comment]
How difficult would it be to create a React 17 compatible version as a temp measure for those of us on React 17? i.e. leave 16 as the main supported version it's time for the whole thing to officially move to be based on React 17?

**Comment by benjamin-andersen** [comment]
1+. This library seems super useful and our team would like to use it in our production application to generate a PDF and download it. We use React 17 however and there aren't many other alternatives that provide a solution to this :(

I think making a separate version for React 17 and one for React 16 would be a good way to temporarily solve this issue as mentioned by @grmatthews.

**Comment by sm3sher** [comment]
Waiting for React 17 support

**Comment by yourjhay** [comment]
Waiting for React 17 Support, Planning to use this lib in our production env

**Comment by bbzitopirulito** [comment]
Any update here? @diegomura

**Comment by muhsalaa** [comment]
Waiting for React 17 Support

**Comment by TD-DO** [comment]
Would also like React 17 support

**Comment by indignant** [comment]
I was curious how other packages are handling this. react-router [is using 17](https://github.com/remix-run/react-router/blob/v6.0.0/package.json#L51) and react-redux [uses both](https://github.com/reduxjs/react-redux/blob/v7.2.6/package.json#L39). For what it's worth, Wes Bos [recommends the same thing](https://twitter.com/wesbos/status/1359173446740172810) as @mquandalle.

As of the end of October Node 16 (NPM 7) is the active LTS; I think this is going to be increasingly problematic as people start switching over and I think it [should be a quick fix](https://github.com/diegomura/react-pdf/issues/1062#issuecomment-836941774). Would be glad to submit a PR with maybe some hints on how I can test.

**Comment by vegai** [comment]
PR above implements what mquandelle suggested on May 10, 2021. Seems to fix the problem in our end at least.

**Comment by dudusohne** [comment]
Still doesn't worked for me

**Comment by karlhorky** [comment]
@diegomura with the React version 18 release just around the corner, what do you think about upgrading to React 17 (dropping support for React 16) and publishing a new major version of `@react-pdf/renderer`?

**Comment by karlhorky** [comment]
One additional detail here is that upgrading to React 17 enables usage of [the new JSX transform](https://reactjs.org/blog/2020/09/22/introducing-the-new-jsx-transform.html), which would allow for leaving out the `import React from 'react'` in every file! 🙌

**Comment by karlhorky** [comment]
Just ran into a problem trying to use `@react-pdf/renderer` with the new transform, as my TypeScript `tsconfig.json` specified the new transform in `compilerOptions`:

```json
{
  "compilerOptions": {
    "jsx": "react-jsx"
  }
}
```

---

To fix this, I needed to:

1. Change `jsx` to `preserve` (see below) to get `@react-pdf/renderer` to work
2. Add the `import React from 'react'` to the file using `@react-pdf/renderer`

```diff
{
  "compilerOptions": {
-    "jsx": "react-jsx"
+    "jsx": "preserve"
  }
}
```

**Comment by karlhorky** [comment]
cc @jeetiss, since it seems that you are also an active maintainer here

**Comment by pascalporedda** [comment]
Gonna +1 the support of React 18.

**Comment by karlhorky** [comment]
`@react-pdf/renderer` is working also with React 18 for us in production - after [our workarounds for React 17](https://github.com/diegomura/react-pdf/issues/1062#issuecomment-1079225111)

**Comment by karlhorky** [comment]
Hm, but it fails with [`@types/react` v18](https://www.npmjs.com/package/@types/react), which was just published:

```
Run yarn tsc
yarn run v1.22.18
$ /home/runner/work/project/node_modules/.bin/tsc
Error: project/generatePdf.tsx(212,6): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: DocumentProps | Readonly<DocumentProps>): Document', gave the following error.
    Type '{ children: Element[]; }' has no properties in common with type 'IntrinsicAttributes & IntrinsicClassAttributes<Document> & Readonly<DocumentProps>'.
  Overload 2 of 2, '(props: DocumentProps, context: any): Document', gave the following error.
    Type '{ children: Element[]; }' has no properties in common with type 'IntrinsicAttributes & IntrinsicClassAttributes<Document> & Readonly<DocumentProps>'.
Error: project/generatePdf.tsx(213,8): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: PageProps | Readonly<PageProps>): Page', gave the following error.
    Type '{ children: Element; size: "A4"; style: { fontFamily: string; fontSize: number; lineHeight: number; paddingTop: number; paddingBottom: number; paddingHorizontal: number; }; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
  Overload 2 of 2, '(props: PageProps, context: any): Page', gave the following error.
    Type '{ children: Element; size: "A4"; style: { fontFamily: string; fontSize: number; lineHeight: number; paddingTop: number; paddingBottom: number; paddingHorizontal: number; }; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
Error: packages/learn.upleveled.io-server/routes/studentAssessments/learningOutcomesPdf.tsx(259,12): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: SVGProps | Readonly<SVGProps>): Svg', gave the following error.
    Type '{ children: Element[]; style: { marginTop: number; marginBottom: number; }; width: string; height: string; viewBox: string; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Svg> & Readonly<SVGProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Svg> & Readonly<SVGProps>'.
  Overload 2 of 2, '(props: SVGProps, context: any): Svg', gave the following error.
    Type '{ children: Element[]; style: { marginTop: number; marginBottom: number; }; width: string; height: string; viewBox: string; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Svg> & Readonly<SVGProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Svg> & Readonly<SVGProps>'.
Error: project/generatePdf.tsx(266,16): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: LinearGradientProps | Readonly<LinearGradientProps>): LinearGradient', gave the following error.
    Type '{ children: Element[]; id: string; x1: string; x2: string; y1: string; y2: string; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<LinearGradient> & Readonly<LinearGradientProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<LinearGradient> & Readonly<LinearGradientProps>'.
  Overload 2 of 2, '(props: LinearGradientProps, context: any): LinearGradient', gave the following error.
    Type '{ children: Element[]; id: string; x1: string; x2: string; y1: string; y2: string; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<LinearGradient> & Readonly<LinearGradientProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<LinearGradient> & Readonly<LinearGradientProps>'.
Error: project/generatePdf.tsx(285,8): error TS2769: No overload matches this call.
  Overload 1 of 2, '(props: PageProps | Readonly<PageProps>): Page', gave the following error.
    Type '{ children: Element; size: "A4"; style: { fontFamily: string; fontSize: number; lineHeight: number; paddingTop: number; paddingBottom: number; paddingHorizontal: number; }; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
  Overload 2 of 2, '(props: PageProps, context: any): Page', gave the following error.
    Type '{ children: Element; size: "A4"; style: { fontFamily: string; fontSize: number; lineHeight: number; paddingTop: number; paddingBottom: number; paddingHorizontal: number; }; }' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
      Property 'children' does not exist on type 'IntrinsicAttributes & IntrinsicClassAttributes<Page> & Readonly<PageProps>'.
error Command failed with exit code 2.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
Error: Process completed with exit code 2.
```

**Comment by karlhorky** [comment]
> Hm, but it fails with [`@types/react` v18](https://www.npmjs.com/package/@types/react), which was just published (because of [the breaking changes](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/56210))

Found some time to take a look at this, and opened a PR to fix the `@react-pdf/renderer` package: https://github.com/diegomura/react-pdf/pull/1798

However, this does not fix the other packages in the monorepo.

cc @diegomura @jeetiss

**Comment by mabreu0** [comment]
In the meantime, while i get more insight about this.. Watchers please drop any light if any!

(This is using **--legacy-peer-deps** alternative with **react-scripts 5.0.1 / react 18**)

Creating an optimized production build...
Failed to compile.

Module not found: Error: Can't resolve 'stream' in '.....**/node_modules/@react-pdf/pdfkit/lib**'
BREAKING CHANGE: webpack < 5 used to include polyfills for node.js core modules by default.
This is no longer the case. Verify if you need this module and configure a polyfill for it.

If you want to include a polyfill, you need to:
	- add a fallback 'resolve.fallback: { "stream": require.resolve("stream-browserify") }'
	- install 'stream-browserify'
If you don't want to include a polyfill, you can use an empty module like this:
	resolve.fallback: { "stream": false }

**Comment by MirKml** [comment]
I think that possible solution is use `react-reconciler` as peerDep for `@react-pdf/renderer` itself.
Renderer itself than can be backward compatible for react16 - as @diegomura wants. It's up to user which version of reconciler will be used.
Reconciler itself has peerDep for correct react major version. 
So then end user has minimal space for error. Maybe npm 7+ with new peerDep auto solving process can install correct version of reconciler. If not, there will be information in install section from package autor :-)

Nowadays, latest major `@react-pdf/renderer` declares compatibility with react 17 via peerDep, but has runtime dependency for old react-reconciler - 0.23, which isn't compatible with react 17 but it works together. Magic from user point of view :-)
When user tries (doesn't matter how :-)) 'correct' react-reconciler 0.26 - version for react 17 - installation is without any warnings but renderer fails in runtime. I think it's weird and has to be fixed.

**Comment by ops1ops** [comment]
Any news? Will the fate of the enzyme overtake this library? Its been 2 years but you still dont support React 17...

**Comment by andfaulkner** [comment]
Still seeking React 17 support at least (if not React 18).

**Comment by jeetiss** [comment]
should be fixed in #2140
